lib/identity_cache/tracking.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module IdentityCache
+  module Tracking
+    TrackedObject = Struct.new(:object, :includes, :caller, :accessed_associations)
+
+    extend self
+
+    def tracked_objects
+      Thread.current[:idc_tracked_objects] ||= {}
+    end
+
+    def reset_tracked_objects
+      Thread.current[:idc_tracked_objects] = {}
+    end
+
+    def track_object(object, includes)
+      return unless object_tracking_enabled
+      locations = caller(1, 20)
+      tracked_objects[object] = TrackedObject.new(object, Array(includes), locations, Set.new)
+    end
+
+    def track_association_accessed(object, association_name)
+      return unless object_tracking_enabled
+      obj = self.tracked_objects[object]
+      obj.accessed_associations << association_name if obj
```

Comment:
If an association is accessed multiple times, won't this make the array much larger.  Seems like this could result in memory usage and CPU overhead issues.

Reply:
oh, this one is a set, rather it is the `includes` that is the array.